### PR TITLE
refactor(ITS): efficient Metadata decoding

### PR DESCRIPTION
**Pull request by re1ro**
[AXE-2516]
* [x] `InterchainTokenService`: efficient `_decodeMetadata` via calldata slices
* [x] `TokenManager`: gas efficient `callContractWithInterchainToken` via `ITS.transmitInterchainTransferWithData` by avoiding `encodePacked` and any conversions between `bytes calldata <-> memory` for  `data`

[AXE-2516]: https://axelarnetwork.atlassian.net/browse/AXE-2516?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: contracts/InterchainTokenService.sol [hunk 1]
```diff
@@ -514,7 +509,32 @@ contract InterchainTokenService is
...
+     */
+    function transmitInterchainTransferWithData(
+        bytes32 tokenId,
+        address sourceAddress,
+        string calldata destinationChain,
+        bytes memory destinationAddress,
+        uint256 amount,
+        uint32 metadataVersion,
+        bytes memory data
+    ) external payable onlyTokenManager(tokenId) whenNotPaused {
```
Comment (deanamiel) [inline contracts/InterchainTokenService.sol:536]: These modifiers are not covered in unit tests, I can add coverage in a separate PR once changes are finalized

**Review by deanamiel — APPROVED** [review]

File: contracts/InterchainTokenService.sol [hunk 2]
```diff
@@ -476,20 +478,13 @@ contract InterchainTokenService is
...
 
-        _transmitInterchainTransfer(
-            tokenId,
-            msg.sender,
-            destinationChain,
-            destinationAddress,
-            amount,
-            abi.encodePacked(LATEST_METADATA_VERSION, data)
-        );
+        _transmitInterchainTransfer(tokenId, msg.sender, destinationChain, destinationAddress, amount, LATEST_METADATA_VERSION, data);
```
Comment (Foivos) [inline contracts/InterchainTokenService.sol:487]: Not sure when LATEST_METADATA_VERSION became a constant, but here we should use a more specific constant )which is the same as LATEST_METADATA_VERSION for now) which would be named something like LATEST_TRANSMIT_INTERCHAIN_TRANSFER_METADATA_VERSION.

File: contracts/token-manager/TokenManager.sol [hunk 3]
```diff
@@ -172,13 +172,14 @@ abstract contract TokenManager is ITokenManager, ITokenManagerType, Operatable,
...
         // slither-disable-next-line var-read-using-this
-        interchainTokenService.transmitInterchainTransfer{ value: msg.value }(
+        interchainTokenService.transmitInterchainTransferWithData{ value: msg.value }(
             this.interchainTokenId(),
             sender,
             destinationChain,
             destinationAddress,
             amount,
-            abi.encodePacked(LATEST_METADATA_VERSION, data)
+            LATEST_METADATA_VERSION,
```
Comment (Foivos) [inline contracts/token-manager/TokenManager.sol:181]: same for this

**Review by milapsheth — APPROVED** [review]

File: contracts/token-manager/TokenManager.sol (hunk 3 above)
Comment (re1ro) [reply contracts/token-manager/TokenManager.sol:181]: Will do in a separate PR

File: contracts/InterchainTokenService.sol (hunk 1 above)
Comment (re1ro) [reply contracts/InterchainTokenService.sol:533]: @deanamiel good observation! The PR is finalized, please add those

Comment (deanamiel) [reply contracts/InterchainTokenService.sol:533]: sounds good, added!

**Pull request merged**